Source/Core/VideoCommon/Resources/ShaderResource.cpp

```diff
@@ -0,0 +1,310 @@
+  return g_gfx->CreateShaderFromSource(ShaderStage::Vertex, source_code.GetBuffer(), nullptr,
+                                       "Custom Vertex Shader");
+}
+}  // namespace
+ShaderResource::ShaderResource(Resource::ResourceContext resource_context,
+                               const GXPipelineUid& pipeline_uid,
+                               const std::string& preprocessor_setting)
+    : Resource(std::move(resource_context)), m_uid(pipeline_uid),
+      m_preprocessor_settings(preprocessor_setting)
+{
+  m_shader_asset = m_resource_context.asset_cache->CreateAsset<RasterSurfaceShaderAsset>(
+      m_resource_context.primary_asset_id, m_resource_context.asset_library, this);
+}
+
+void ShaderResource::SetHostConfig(const ShaderHostConfig& host_config)
+{
+  m_shader_host_config.bits = host_config.bits;
+}
+
+void ShaderResource::MarkAsPending()
+{
+  m_resource_context.asset_cache->MarkAssetPending(m_shader_asset);
+}
+
+void ShaderResource::MarkAsActive()
+{
+  m_resource_context.asset_cache->MarkAssetActive(m_shader_asset);
+}
+
+AbstractShader* ShaderResource::Data::GetVertexShader() const
+{
+  if (!m_vertex_shader)
+    return nullptr;
+  return m_vertex_shader.get();
+}
+
+AbstractShader* ShaderResource::Data::GetPixelShader() const
+{
+  if (!m_pixel_shader)
+    return nullptr;
+  return m_pixel_shader.get();
+}
+
+AbstractShader* ShaderResource::Data::GetGeometryShader() const
+{
+  if (!m_geometry_shader)
+    return nullptr;
+  return m_geometry_shader.get();
+}
+
+bool ShaderResource::Data::IsCompiled() const
+{
+  return m_vertex_shader && m_pixel_shader && (!m_needs_geometry_shader || m_geometry_shader);
+}
+
+AbstractTextureType ShaderResource::Data::GetTextureType(std::size_t index)
+{
+  // If the data doesn't exist, just pick one...
+  if (!m_shader_data || index >= m_shader_data->samplers.size()) [[unlikely]]
+    return AbstractTextureType::Texture_2D;
+
+  return m_shader_data->samplers[index].type;
+}
+
+void ShaderResource::ResetData()
+{
+  m_load_data = std::make_shared<Data>();
+  m_processing_load_data = false;
+}
+
+Resource::TaskComplete ShaderResource::CollectPrimaryData()
+{
+  const auto shader_data = m_shader_asset->GetData();
+  if (!shader_data) [[unlikely]]
+  {
+    return Resource::TaskComplete::No;
+  }
+  m_load_data->m_shader_data = shader_data;
+
+  return Resource::TaskComplete::Yes;
+}
+
+Resource::TaskComplete ShaderResource::ProcessData()
+{
+  class WorkItem final : public VideoCommon::AsyncShaderCompiler::WorkItem
+  {
+  public:
+    WorkItem(std::shared_ptr<ShaderResource::Data> resource_data, VideoCommon::GXPipelineUid* uid,
+             u32 shader_bits, std::string_view preprocessor_settings)
+        : m_resource_data(std::move(resource_data)), m_uid(uid), m_shader_bits(shader_bits),
+          m_preprocessor_settings(preprocessor_settings)
+    {
+    }
+
+    bool Compile() override
+    {
+      ShaderHostConfig shader_host_config;
+      shader_host_config.bits = m_shader_bits;
```

Comment:
```suggestion
      ShaderHostConfig shader_host_config{.bits = m_shader_bits};
```
Eliminates a warning for `shader_host_config` being uninitialized.